### PR TITLE
Support switching control mode on the fly

### LINKED ISSUE #6: Support switching control pattern on-the-fly.
As we are adding more control patterns to autoscaler like #5, it would be great to also support switching control pattern without restarting the autoscaler.

Currently the `control mode` is passed to autoscaler as an input argument. One thought to achieve this is also parse out the `control mode` from ConfigMap.

### PR DESCRIPTION
**Pull request by MrHohn**
Fixed #6.

Parses out the specified control mode from configMap in loop and ensures corresponding controller on changes.

Adds unit test cases for switching control pattern and linear controller in mocking server test.

@bowei

### CODE REVIEW

File: pkg/autoscaler/controller/plugin/plugin.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"fmt"
+
+	"github.com/kubernetes-incubator/cluster-proportional-autoscaler/pkg/autoscaler/controller"
+	"github.com/kubernetes-incubator/cluster-proportional-autoscaler/pkg/autoscaler/controller/laddercontroller"
+	"github.com/kubernetes-incubator/cluster-proportional-autoscaler/pkg/autoscaler/controller/linearcontroller"
+	"github.com/kubernetes-incubator/cluster-proportional-autoscaler/pkg/autoscaler/k8sclient"
+)
+
+func GetControlMode(configMap *k8sclient.ConfigMap) (string, error) {
+	mode := ""
```
Comment (bowei) [inline pkg/autoscaler/controller/plugin/plugin.go]: why not 

```go
func GetController(configMap) (controller.Controller, error) {
  if len(configMap.Data) != 1 {
    return nil, fmt.Errorf("invalid ...)
  }

  for m, v := range configMap.Data {
    switch m {
    case laddercontroller...
    } 
}
```

and return the controller directly?

File: cmd/cluster-proportional-autoscaler/autoscaler.go [hunk 2]
```diff
@@ -46,7 +46,7 @@ func main() {
 		os.Exit(1)
 	}
 
-	glog.V(0).Infof("Scaling Namespace: %s, Target: %s, Mode: %v\n", config.Namespace, config.Target, config.Mode)
+	glog.V(0).Infof("Scaling Namespace: %s, Target: %s\n", config.Namespace, config.Target)
```
Comment (MrHohn) [inline cmd/cluster-proportional-autoscaler/autoscaler.go]: Done. Removed other unnecessary `\n`s as well.

File: pkg/autoscaler/autoscaler_server.go [hunk 3]
```diff
@@ -108,6 +96,19 @@ func (s *AutoScaler) pollAPIServer() {
 
 	// Only sync updated ConfigMap
 	if configMap.Version != s.controller.GetParamsVersion() {
+		// Ensure corresponding controller type
+		mode, err := plugin.GetControlMode(configMap)
+		if err != nil {
+			glog.Errorf("error getting control mode: %v\n", err)
```
Comment (MrHohn) [inline pkg/autoscaler/autoscaler_server.go]: Done.

File: pkg/autoscaler/autoscaler_server.go [hunk 4]
```diff
@@ -108,6 +96,19 @@ func (s *AutoScaler) pollAPIServer() {
...
+			return
+		}
+		if s.controller == nil || mode != s.controller.GetControllerType() {
+			s.controller, err = plugin.GetController(mode)
+			if err != nil {
+				glog.Errorf("error getting controller: %v\n", err)
+				return
+			}
+		}
```
Comment (MrHohn) [inline pkg/autoscaler/autoscaler_server.go]: Adds logging in EnsureController().

File: pkg/autoscaler/controller/plugin/plugin.go (hunk 1 above)
Comment (MrHohn) [inline pkg/autoscaler/controller/plugin/plugin.go]: Make sense. Combined these two parts as EnsureController().

**Comment by MrHohn** [comment]
Thanks for reviewing. The new commit also adds unit tests for plugin package.

**Pull request merged**